naga/src/back/glsl/mod.rs

```diff
@@ -3463,12 +3468,66 @@ impl<'a, W: Write> Writer<'a, W> {
+                            Mf::Pack4xI8 => true,
+                            Mf::Pack4xU8 => false,
+                            _ => unreachable!(),
+                        };
+                        let const_suffix = if was_signed { "" } else { "u" };
+                        if was_signed {
+                            write!(self.out, "uint(")?;
+                        }
+                        write!(self.out, "(")?;
+                        self.write_expr(arg, ctx)?;
+                        write!(self.out, "[0] & 0xFF{const_suffix}) | ((")?;
+                        self.write_expr(arg, ctx)?;
+                        write!(self.out, "[1] & 0xFF{const_suffix}) << 8) | ((")?;
+                        self.write_expr(arg, ctx)?;
+                        write!(self.out, "[2] & 0xFF{const_suffix}) << 16) | ((")?;
+                        self.write_expr(arg, ctx)?;
+                        write!(self.out, "[3] & 0xFF{const_suffix}) << 24)")?;
+                        if was_signed {
+                            write!(self.out, ")")?;
+                        }
+
+                        return Ok(());
+                    }
                     // data unpacking
                     Mf::Unpack4x8snorm => "unpackSnorm4x8",
                     Mf::Unpack4x8unorm => "unpackUnorm4x8",
                     Mf::Unpack2x16snorm => "unpackSnorm2x16",
                     Mf::Unpack2x16unorm => "unpackUnorm2x16",
                     Mf::Unpack2x16float => "unpackHalf2x16",
+                    fun @ (Mf::Unpack4xI8 | Mf::Unpack4xU8) => {
+                        let sign_prefix = match fun {
+                            Mf::Unpack4xI8 => 'i',
+                            Mf::Unpack4xU8 => 'u',
+                            _ => unreachable!(),
+                        };
+                        write!(self.out, "{sign_prefix}vec4(")?;
+                        for i in 0..4 {
+                            write!(self.out, "bitfieldExtract(")?;
+                            // Since bitfieldExtract only sign extends if the value is signed, this
+                            // cast is needed
+                            match fun {
+                                Mf::Unpack4xI8 => {
+                                    write!(self.out, "int(")?;
+                                    self.write_expr(arg, ctx)?;
+                                    write!(self.out, ")")?;
+                                }
+                                Mf::Unpack4xU8 => self.write_expr(arg, ctx)?,
+                                _ => unreachable!(),
+                            };
+                            write!(self.out, ", {}, 8)", i * 8)?;
+                            if i != 3 {
+                                write!(self.out, ", ")?;
+                            }
+                        }
+                        write!(self.out, ")")?;
+
+                        return Ok(());
+                    }
```

Comment:
Can we generate the same code as we do for the MSL and HLSL backends?

Reply:
Yeah, it can be done, but I assume that using bitfieldExtract is cleaner than using bitshifts to emulate sign extension. Also, does stuff in less instructions. If it's important to keep all of the implementations as similar as possible, I can rewrite them.

Reply:
I usually lean towards consistency. I thought `bitfieldExtract` will get lowered into bitshifts anyway but it seems at least nvidia and amd have instructions in their ISA that correspond to `bitfieldExtract`. Even so I would imagine the hardware is more complex for `bitfieldExtract` since it's a more complex operation.

I propose we go with bitshifts everywhere unless `bitfieldExtract` is faster.

Reply:
As far as I know bfi is faster, multiple shifts require multiple instructions, and bfi and shifts should work at the same rate.

Reply:
Still, we need to keep in mind that less instructions doesn't always mean that it's going to be faster. Maybe one shift uses one microinstruction and one bitfield operation uses 6, if not optimized on hardware level. So I agree that this choice requires benchmarking. 
The problem I have right now is that I don't have that much hardware to test on, so it may be good on my system, but worse on many others.
Also, in the future, bitfield operations may become faster if they are slow now, but that seems more like a problem for future people.

Reply:
Let's keep things as they are for now since it's not clear what is best.